src/sentry/filters/preprocess_hashes.py

```diff
@@ -11,3 +16,153 @@
+        # coming from an HTML page or it could be minified/unparseable
+        # code, so lets defer to other lesser heuristics (like lineno)
+        can_use_context = False
+    else:
+        can_use_context = True
+
+    if can_use_context:
+        output.append(context_line)
+    elif not output:
+        # If we were unable to achieve any context at this point
+        # (likely due to a bad JavaScript error) we should just
+        # bail on recording this frame
+        return output
+    elif frame.get('symbol'):
+        output.append(frame.get('symbol'))
+    elif frame.get('function'):
+        if is_unhashable_function(frame['function']):
+            output.append('<function>')
+        else:
+            output.append(remove_function_outliers(frame['function']))
+    elif frame.get('lineno') is not None and frame.get('colno') is not None:
+        output.extend([frame['lineno'], frame['colno']])
+    return output
+
+
+def get_stacktrace_hash(platform, stacktrace):
+    frames = stacktrace['frames']
+
+    stack_invalid = is_invalid_stack(frames)
+    if stack_invalid:
+        return []
+
+    output = []
+    for frame in frames:
+        output.extend(get_frame_hash(platform, frame))
+
+    return output
+
+
+def get_exception_hash(platform, exception):
+    output = []
+    for value in exception['values']:
+        if not value['stacktrace']:
+            continue
+        stack_hash = get_stacktrace_hash(platform, value['stacktrace'])
+        if stack_hash:
+            output.extend(stack_hash)
+            output.append(value.get('type'))
+
+    if not output:
+        for value in exception['values']:
+            output.extend([v for v in [value.get('type'), value.get('value')] if v])
+
+    return output
+
+
+def get_preprocess_defaults(data):
+    stacktrace = data.get('sentry.interfaces.Stacktrace')
+    exception = data.get('sentry.interfaces.Exception')
+    template = data.get('sentry.interfaces.Template')
+
+    output = []
+
+    if exception is not None:
+        result = get_exception_hash(data['platform'], exception)
+        if result:
+            output.append(result)
+
+    if stacktrace is not None and not output:
+        result = get_stacktrace_hash(data['platform'], stacktrace)
+        if result:
+            output.append(result)
+
+    if template is not None and not output:
+        # i think data should be validated at this point?
+        output.append([template['filename'], template['context_line']])
+
+    if not output and data.get('message'):
+        output = [data['message']]
+
+    return output
+
+
+def get_preprocess_hashes(data):
+    from sentry.coreapi import LazyData
+    if isinstance(data, LazyData):
+        data = dict(data.items())
+
+    fingerprint = data.get('fingerprint')
+    checksum = data.get('checksum')
+    output = get_preprocess_defaults(data)
+
+    if fingerprint:
+        default_values = set(['{{ default }}', '{{default}}'])
```

Comment:
We should make this a constant value defined at module level and turn it into a `frozenset` instead of defining it for a third(!) time:

```
src/sentry/event_manager.py:102:    default_values = set(['{{ default }}', '{{default}}'])
src/sentry/event_manager.py:122:    default_values = set(['{{ default }}', '{{default}}'])
```